README.md
Cabot Hipchat Plugin
=====

This is an alert plugin for the cabot service monitoring tool. It allows you to alert users by their user handle in a hipchat room.

## Installation

Install using pip

    pip install cabot-alert-hipchat

Edit `conf/production.env` in your Cabot clone to include the plugin:

    CABOT_PLUGINS_ENABLED=cabot_alert_hipchat...,<other plugins>

## Configuration

The plugin requires the following environment variables to be set:

    HIPCHAT_ALERT_ROOM=<room_id>
    HIPCHAT_API_KEY=your_hipchat_api_key
    
For self-hosted hipchat deployments you can also specify

    HIPCHAT_DOMAIN=api.my-hipch.at
    
Previous versions used `HIPCHAT_URL`. This is now deprecated.

If `HIPCHAT_URL` is set and `HIPCHAT_DOMAIN` isn't, the domain-name will be parsed from `HIPCHAT_URL`


CHANGES
Version 2.0.3
-------------

* Add serialize method

Version 2.0.2
-------------

* Raise exception in both HIPCHAT_URL and HIPCHAT_DOMAIN are set.
  - Currently there is a bug where nothing happens when both are set

Version 2.0.1
-------------

* Make it backwards compatible
  - If HIPCHAT_URL exists, v1 api will be used, otherwise v2 will be used

Version 2.0.0
-------------

* Switch to using hipchat API v2
  - Now uses the HIPCHAT_DOMAIN for custom hipchat deployments
  - Will extract the domain from HIPCHAT_URL if it exists (for backwards compatibility)


cabot_alert_hipchat/tests.py
from django.contrib.auth.models import User
from cabot.cabotapp.tests.tests_basic import LocalTestCase
from cabot.plugins.models import AlertPluginModel
from mock import Mock, patch

from cabot.cabotapp.models import Service
from cabot_alert_hipchat import plugin

class TestHipchatAlerts(LocalTestCase):
    def setUp(self):
        super(TestHipchatAlerts, self).setUp()

        self.hipchat_plugin, created = AlertPluginModel.objects.get_or_create(
                slug='cabot_alert_hipchat')

        u = User.objects.get(pk=self.user.pk)
        u.cabot_alert_hipchat_settings.hipchat_alias = 'test_user_hipchat_alias'
        
        self.service.users_to_notify.add(self.user)
        self.service.alerts.add(self.hipchat_plugin)
        self.service.save()
        self.service.update_status()

    def test_users_to_notify(self):
        self.assertEqual(self.service.users_to_notify.all().count(), 1)
        self.assertEqual(self.service.users_to_notify.get(pk=1).username, self.user.username)

    @patch('cabot_alert_hipchat.plugin.HipchatAlertPlugin._send_hipchat_alert')
    def test_normal_alert(self, fake_hipchat_alert):
        self.service.overall_status = Service.PASSING_STATUS
        self.service.old_overall_status = Service.ERROR_STATUS
        self.service.save()
        self.service.alert()
        fake_hipchat_alert.assert_called_with(u'Service Service is back to normal: http://localhost/service/1/. @test_user_hipchat_alias', color='green', sender='Cabot/Service')

    @patch('cabot_alert_hipchat.plugin.HipchatAlertPlugin._send_hipchat_alert')
    def test_failure_alert(self, fake_hipchat_alert):
        # Most recent failed
        self.service.overall_status = Service.CALCULATED_FAILING_STATUS
        self.service.old_overall_status = Service.PASSING_STATUS
        self.service.save()
        self.service.alert()
        fake_hipchat_alert.assert_called_with(u'Service Service reporting failing status: http://localhost/service/1/. Checks failing: @test_user_hipchat_alias', color='red', sender='Cabot/Service')



cabot_alert_hipchat/tests/test_hipchat.py
from cabot.cabotapp.tests.tests_basic import LocalTestCase
from mock import Mock, patch

from cabot.cabotapp.models import UserProfile, Service
from cabot_alert_hipchat import models
from cabot.cabotapp.alert import update_alert_plugins

class TestHipchatAlerts(LocalTestCase):
    def setUp(self):
        super(TestHipchatAlerts, self).setUp()

        self.hipchat_user_data = models.HipchatAlertUserData.objects.create(
            hipchat_alias = "test_user_hipchat_alias",
            user = self.user.profile,
            title=models.HipchatAlertUserData.name,
            )
        self.hipchat_user_data.save()

        self.service.users_to_notify.add(self.user)

        update_alert_plugins()
        self.hipchat_plugin = models.HipchatAlert.objects.get(title=models.HipchatAlert.name)
        self.service.alerts.add(self.hipchat_plugin)
        self.service.save()
        self.service.update_status()

    def test_users_to_notify(self):
        self.assertEqual(self.service.users_to_notify.all().count(), 1)
        self.assertEqual(self.service.users_to_notify.get(pk=1).username, self.user.username)

    @patch('cabot_alert_hipchat.models.HipchatAlert._send_hipchat_alert')
    def test_normal_alert(self, fake_hipchat_alert):
        self.service.overall_status = Service.PASSING_STATUS
        self.service.old_overall_status = Service.ERROR_STATUS
        self.service.save()
        self.service.alert()
        fake_hipchat_alert.assert_called_with(u'Service Service is back to normal: http://localhost/service/1/.  @test_user_hipchat_alias', color='green', sender='Cabot/Service')

    @patch('cabot_alert_hipchat.models.HipchatAlert._send_hipchat_alert')
    def test_failure_alert(self, fake_hipchat_alert):
        # Most recent failed
        self.service.overall_status = Service.CALCULATED_FAILING_STATUS
        self.service.old_overall_status = Service.PASSING_STATUS
        self.service.save()
        self.service.alert()
        fake_hipchat_alert.assert_called_with(u'Service Service reporting failing status: http://localhost/service/1/. Checks failing:  @test_user_hipchat_alias', color='red', sender='Cabot/Service')


cabot_alert_hipchat/tests/hipchat_tests.py
from cabot.cabotapp.tests.tests_basic import LocalTestCase
from mock import Mock, patch

from cabot.cabotapp.models import UserProfile, Service
from cabot_alert_hipchat import models
from cabot.cabotapp.alert import update_alert_plugins

class TestHipchatAlerts(LocalTestCase):
    def setUp(self):
        super(TestHipchatAlerts, self).setUp()

        self.user_profile = UserProfile(user=self.user)
        self.user_profile.save()
        self.hipchat_user_data = models.HipchatAlertUserData.objects.create(
            hipchat_alias = "test_user_hipchat_alias",
            user = self.user_profile,
            title=models.HipchatAlertUserData.name,
            )
        self.hipchat_user_data.save()

        self.service.users_to_notify.add(self.user)

        update_alert_plugins()
        self.hipchat_plugin = models.HipchatAlert.objects.get(title=models.HipchatAlert.name)
        self.service.alerts.add(self.hipchat_plugin)
        self.service.save()
        self.service.update_status()

    def test_users_to_notify(self):
        self.assertEqual(self.service.users_to_notify.all().count(), 1)
        self.assertEqual(self.service.users_to_notify.get(pk=1).username, self.user.username)

    @patch('cabot_alert_hipchat.models.HipchatAlert._send_hipchat_alert')
    def test_normal_alert(self, fake_hipchat_alert):
        self.service.overall_status = Service.PASSING_STATUS
        self.service.old_overall_status = Service.ERROR_STATUS
        self.service.save()
        self.service.alert()
        fake_hipchat_alert.assert_called_with(u'Service Service is back to normal: http://localhost/service/1/. @test_user_hipchat_alias', color='green', sender='Cabot/Service')


cabot_alert_hipchat/plugin.py
from django.conf import settings
from django.core.urlresolvers import reverse
from cabot.plugins.models import AlertPlugin
from django import forms
from os import environ as env
import requests


class HipchatUserSettingsForm(forms.Form):
    hipchat_alias = forms.CharField(max_length=64)

class HipchatAlertPlugin(AlertPlugin):
    name = "Hipchat"
    slug = "cabot_alert_hipchat"
    author = "Jonathan Balls"
    version = "0.0.1"
    font_icon = "fa fa-comment"

    plugin_variables = [
        'HIPCHAT_ALERT_ROOM',
        'HIPCHAT_API_KEY'
    ]

    user_config_form = HipchatUserSettingsForm

    def send_alert(self, service, users, duty_officers):

        # Check whether to alert users (by appending their names to the end of
        # the message).
        alert = True
        if service.overall_status == service.WARNING_STATUS:
            alert = False  # Don't alert at all for WARNING
        if service.overall_status == service.ERROR_STATUS:
            if service.old_overall_status in (service.ERROR_STATUS, service.ERROR_STATUS):
                alert = False  # Don't alert repeatedly for ERROR
        if service.overall_status == service.PASSING_STATUS:
            color = 'green'
            if service.old_overall_status == service.WARNING_STATUS:
                alert = False  # Don't alert for recovery from WARNING status
        else:
            color = 'red'

        message = service.get_status_message()

        if alert:
            #users = list(users) + list(duty_officers)
            hipchat_aliases = [u.cabot_alert_hipchat_settings.hipchat_alias for u in users]
            for alias in hipchat_aliases:
                message = '{} @{}'.format(message, alias)

        self._send_hipchat_alert(
            message,
            color=color,
            sender='Cabot/%s' % service.name
        )

    def _send_hipchat_alert(self, message, color='green', sender='Cabot'):

        room = env.get('HIPCHAT_ALERT_ROOM')
        api_key = env.get('HIPCHAT_API_KEY')
        url = env.get('HIPCHAT_URL')

        resp = requests.post(url + '?auth_token=' + api_key, data={
            'room_id': room,
            'from': sender[:15],
            'message': message,
            'notify': 1,
            'color': color,
            'message_format': 'text',
	})



cabot_alert_hipchat/migrations/0001_initial.py
# -*- coding: utf-8 -*-
from __future__ import unicode_literals

from django.db import models, migrations


class Migration(migrations.Migration):

    dependencies = [
        ('cabotapp', '0001_initial'),
    ]

    operations = [
        migrations.CreateModel(
            name='HipchatAlert',
            fields=[
                ('alertplugin_ptr', models.OneToOneField(parent_link=True, auto_created=True, primary_key=True, serialize=False, to='cabotapp.AlertPlugin')),
            ],
            options={
                'abstract': False,
            },
            bases=('cabotapp.alertplugin',),
        ),
        migrations.CreateModel(
            name='HipchatAlertUserData',
            fields=[
                ('alertpluginuserdata_ptr', models.OneToOneField(parent_link=True, auto_created=True, primary_key=True, serialize=False, to='cabotapp.AlertPluginUserData')),
                ('hipchat_alias', models.CharField(max_length=50, blank=True)),
            ],
            options={
                'abstract': False,
            },
            bases=('cabotapp.alertpluginuserdata',),
        ),
    ]

cabot_alert_hipchat/models.py
from django.db import models
from cabot.cabotapp.alert import AlertPlugin, AlertPluginUserData

from os import environ as env

from django.conf import settings
from django.core.urlresolvers import reverse
from django.template import Context, Template

import logging
import requests
from urlparse import urlparse


logger = logging.getLogger(__name__)

hipchat_template = "Service {{ service.name }} {% if service.overall_status == service.PASSING_STATUS %}is back to normal{% else %}reporting {{ service.overall_status }} status{% endif %}: {{ scheme }}://{{ host }}{% url 'service' pk=service.id %}. {% if service.overall_status != service.PASSING_STATUS %}Checks failing: {% for check in service.all_failing_checks %}{% if check.check_category == 'Jenkins check' %}{% if check.last_result.error %} {{ check.name }} ({{ check.last_result.error|safe }}) {{jenkins_api}}job/{{ check.name }}/{{ check.last_result.job_number }}/console{% else %} {{ check.name }} {{jenkins_api}}/job/{{ check.name }}/{{check.last_result.job_number}}/console {% endif %}{% else %} {{ check.name }} {% if check.last_result.error %} ({{ check.last_result.error|safe }}){% endif %}{% endif %}{% endfor %}{% endif %}{% if alert %}{% for alias in users %} @{{ alias }}{% endfor %}{% endif %}"
hipchat_update_template = '{{ service.unexpired_acknowledgement.user.email }} is working on service {{ service.name }} (status {{ service.overall_status }}) - acknowledged @ {{ service.unexpired_acknowledgement.time|date:"H:i" }}'

# This provides the hipchat alias for each user. Each object corresponds to a User
class HipchatAlert(AlertPlugin):
    name = "Hipchat"
    author = "Jonathan Balls"

    def send_alert(self, service, users, duty_officers):
        alert = True
        hipchat_aliases = []
        users = list(users) + list(duty_officers)

        hipchat_aliases = [u.hipchat_alias for u in HipchatAlertUserData.objects.filter(user__user__in=users)]

        if service.overall_status == service.WARNING_STATUS:
            alert = False  # Don't alert at all for WARNING
        if service.overall_status == service.ERROR_STATUS:
            if service.old_overall_status in (service.ERROR_STATUS, service.ERROR_STATUS):
                alert = False  # Don't alert repeatedly for ERROR
        if service.overall_status == service.PASSING_STATUS:
            color = 'green'
            if service.old_overall_status == service.WARNING_STATUS:
                alert = False  # Don't alert for recovery from WARNING status
        else:
            color = 'red'

        c = Context({
            'service': service,
            'users': hipchat_aliases,
            'host': settings.WWW_HTTP_HOST,
            'scheme': settings.WWW_SCHEME,
            'alert': alert,
            'jenkins_api': settings.JENKINS_API,
        })
        message = Template(hipchat_template).render(c)
        self._send_hipchat_alert(
            message,
            color=color,
            sender='Cabot/%s' % service.name
        )

    def send_alert_update(self, service, users, duty_officers):

        c = Context({
            'service': service,
            'host': settings.WWW_HTTP_HOST,
            'scheme': settings.WWW_SCHEME,
            'alert': False,
        })
        message = Template(hipchat_update_template).render(c)
        self._send_hipchat_alert(
            message,
            color='yellow',
            sender='Cabot/%s' % service.name
        )

    def _send_hipchat_alert(self, message, color='green', sender='Cabot'):

        room = env.get('HIPCHAT_ALERT_ROOM')
        api_key = env.get('HIPCHAT_API_KEY')
        domain = env.get('HIPCHAT_DOMAIN', 'api.hipchat.com')

        # Backwards compatibility
        if env.get('HIPCHAT_URL'):
            logger.warn('HIPCHAT_URL is deprecated. Please use HIPCHAT_DOMAIN instead with Hipchat API v2.')

            if env.get('HIPCHAT_DOMAIN'):
                logger.error('HIPCHAT_URL and HIPCHAT_DOMAIN cannot both be set.')
                raise Exception('HIPCHAT_URL and HIPCHAT_DOMAIN cannot both be set, please remove HIPCHAT_URL from your config')
            else:
                resp = requests.post(env.get('HIPCHAT_URL') + '?auth_token=' + api_key, data={
                    'room_id': room,
                    'from': sender[:15],
                    'message': message,
                    'notify': 1,
                    'color': color,
                    'message_format': 'text',
                })
        else:
            url = "https://{domain}/v2/room/{room}/notification?auth_token={api_key}".format(
                domain=domain,
                room=room,
                api_key=api_key
            )

            resp = requests.post(url, data={
                'message': message,
                'notify': 'true',
                'color': color,
                'message_format': 'text',
            })

class HipchatAlertUserData(AlertPluginUserData):
    name = "Hipchat Plugin"
    hipchat_alias = models.CharField(max_length=50, blank=True)

    def serialize(self):
        return {
            "hipchat_alias": self.hipchat_alias
        }

